dynamics/src/include/DynamicsKernel.hpp

```diff
@@ -0,0 +1,334 @@
+/*!
+ * @file DynamicsKernel.hpp
+ *
+ * @date 17 Feb 2023
+ * @author Tim Spain <timothy.spain@nersc.no>
+ * @author Piotr Minakowski <piotr.minakowski@ovgu.de>
+ */
+
+#ifndef DYNAMICSKERNEL_HPP
+#define DYNAMICSKERNEL_HPP
+
+#include "Interpolations.hpp"
+#include "ParametricMesh.hpp"
+#include "ParametricTools.hpp"
+#include "DGTransport.hpp"
+#include "cgParametricMomentum.hpp"
+
+#include "cgVector.hpp"
+#include "dgVector.hpp"
+#include "dgLimit.hpp"
+#include "dgVisu.hpp"
+#include "Tools.hpp"
+
+
+#include "CGModelArray.hpp"
+#include "DGModelArray.hpp"
+#include "include/gridNames.hpp"
+#include "include/Time.hpp"
+
+
+#include <string>
+#include <unordered_map>
+
+namespace Nextsim {
+
+template <int CGdegree, int DGadvection> class DynamicsKernel {
+public:
+    
+    void initialisation() {
+
+        //! Define the spatial mesh
+        smesh = new Nextsim::ParametricMesh(Nextsim::CARTESIAN);
+        // FIXME integrate the creation of the smesh based on restart file
+        //smesh->readmesh("init_topaz128x128.smesh"); // file temporary committed
+        smesh->readmesh("25km_NH.smesh");
+        
+        // output land mask
+        Nextsim::DGVector<1> landmask(*smesh);
+        for (size_t i=0;i<smesh->nelements;++i)
+            landmask(i,0) = smesh->landmask[i];
+        Nextsim::VTK::write_dg<1>("landmask", 0, landmask, *smesh);
+
+
+        // output boundary info
+        Nextsim::DGVector<1> boundary(*smesh);
+        for (size_t j=0;j<4;++j)
+            for (size_t i=0;i<smesh->dirichlet[j].size();++i)
+            boundary(smesh->dirichlet[j][i],0) = 1+j;
+        Nextsim::VTK::write_dg<1>("boundary", 0, boundary, *smesh);
+
+
+        //! Initialize transport
+        dgtransport = new Nextsim::DGTransport<DGadvection>(*smesh);
+        dgtransport->settimesteppingscheme("rk2");
+
+        //! Initialize momentum
+        momentum = new Nextsim::CGParametricMomentum<CGdegree>(*smesh);
+
+
+        //! initialize Forcing 
+        Nextsim::Interpolations::Function2CG(*smesh, momentum->GetOceanx(), OceanX());
+        Nextsim::Interpolations::Function2CG(*smesh, momentum->GetOceany(), OceanY());
+
+        AtmForcingX = new AtmX ;
+        AtmForcingY = new AtmY ;
+
+        AtmForcingX->settime(0.0);
+        AtmForcingY->settime(0.0);
+        Nextsim::Interpolations::Function2CG(*smesh, momentum->GetAtmx(), *AtmForcingX);
+        Nextsim::Interpolations::Function2CG(*smesh, momentum->GetAtmy(), *AtmForcingY);
+
+
+        //resize CG and DG vectors
+        hice.resize_by_mesh(*smesh);
+        cice.resize_by_mesh(*smesh);
+
+        u.resize_by_mesh(*smesh);
+        v.resize_by_mesh(*smesh);
+    }
+
+
+    /*
+        Temrorary Ocen and Wind Field
+    */
+
+    class OceanX : public Nextsim::Interpolations::Function {
+    public:
+        double operator()(double x, double y) const
+        {
+            return  0.01 * (2.0 * y / ( 121. * 25000.) - 1.0);
```

Comment:
There are some magic numbers here which are probably better extracted as constants, particular 121/154 and 25000 (here and line 107, 121,122,142,143)

Reply:
Very much so!
However all of this code is removed [here](https://github.com/nextsimdg/nextsimdg/pull/327/files#diff-906d264d24ee76696155c8cabfaeccccf0fbe08736c467b65e7ebb705f534015) as part of the development that gets the main model and the dynamics talking to each other correctly. After merging this PR, the dynamics still has no effect on the model, so these awful hard coded constants can be ignored for now.

Reply:
Gotcha.